modules/core/core-content/src/main/java/com/enonic/xp/core/impl/content/DeletedEventSyncCommand.java

```diff
@@ -21,37 +33,92 @@ public static Builder create()
+        doSync( params.getContents() );
     }
 
-    private void doSync( final ContentToSync content )
+    private void doSync( final List<ContentToSync> contents )
     {
-        content.getTargetContext().runWith( () -> {
+        final Set<ContentId> fullIds =
+            contents.stream().map( contentToSync -> contentToSync.getTargetContent().getId() ).collect( Collectors.toSet() );
+
+        final Set<ContentToSync> roots = getRoots( contents );
+
+        roots.forEach( content -> content.getTargetContext().runWith( () -> {
 
             if ( isToSyncDelete( content.getTargetContent() ) )
             {
-                if ( needToDelete( content ) )
+                if ( needToDelete( content, fullIds ) )
                 {
                     final DeleteContentParams deleteParams =
                         DeleteContentParams.create().contentPath( content.getTargetContent().getPath() ).build();
 
                     contentService.deleteWithoutFetch( deleteParams );
                 }
             }
-        } );
+        } ) );
+    }
+
+    private Set<ContentToSync> getRoots( final List<ContentToSync> contents )
+    {
+        final Set<ContentPath> paths =
+            contents.stream().map( contentToSync -> contentToSync.getTargetContent().getPath() ).collect( Collectors.toSet() );
+
+        final Set<ContentPath> rootPaths =
+            paths.stream().filter( path -> !paths.contains( path.getParentPath() ) ).collect( Collectors.toSet() );
+
+        return contents.stream()
+            .filter( content -> rootPaths.contains( content.getTargetContent().getPath() ) )
+            .collect( Collectors.toSet() );
     }
 
     private boolean isToSyncDelete( final Content targetContent )
     {
         return targetContent.getInherit().contains( ContentInheritType.CONTENT );
     }
 
-    private boolean needToDelete( final ContentToSync content )
+    private boolean needToDelete( final ContentToSync content, final Set<ContentId> idsToRemove )
+    {
+        return content.getSourceContent() == null && removedInSource( content ) && hasNoInboundDependencies( content, idsToRemove ) &&
+            hasNoChildren( content, idsToRemove );
+    }
+
+    private boolean removedInSource( final ContentToSync contentToSync )
+    {
+        return contentToSync.getSourceContext().callWith( () -> !contentService.contentExists( contentToSync.getTargetContent().getId() ) );
+    }
+
+    private boolean hasNoChildren( final ContentToSync contentToSync, final Set<ContentId> idsToRemove )
+    {
+        return !contentToSync.getTargetContent().hasChildren() ||
+            getAllChildren( contentToSync.getTargetContent().getId() ).stream().allMatch( idsToRemove::contains );
+    }
+
+    private boolean hasNoInboundDependencies( final ContentToSync contentToSync, final Set<ContentId> idsToRemove )
+    {
+        return getInboundDependencies( contentToSync.getTargetContent().getId() ).stream().allMatch( idsToRemove::contains );
```

Comment:
Optimize - move to query

Reply:
'idsToRemove' conversation could be really big